app/src/main/java/de/westnordost/streetcomplete/data/osm/split/SplitWayUpload.kt

```diff
@@ -0,0 +1,258 @@
+            // unsolvable conflict if node has been moved
+            if(old.position.latitude != new.position.latitude || old.position.longitude != new.position.longitude)
+                throw ConflictException("Node #${old.id} has been moved and the conflict cannot be solved automatically")
+        }
+    }
+
+    private fun splitWayAtIndices(originalWay: Way, splitIndices: List<Int>): List<Element> {
+        val newWays = createSplitWays(originalWay, splitIndices)
+        val updatedRelations = updateRelations(originalWay, newWays)
+        return newWays + updatedRelations
+    }
+
+    /** Returns the elements that have been changed */
+    private fun createSplitWays(originalWay: Way, splitIndices: List<Int>): List<Way> {
+        val nodesChunks = originalWay.nodeIds.splitIntoChunks(splitIndices)
+        /* Handle circular ways specially: If you split at a circular way at two nodes, you just
+           want to split it at these points, not also at the former endpoint. So if the last node is
+           the same first node, join the last and the first way chunk. (copied from JOSM) */
+        if (nodesChunks.size > 1 && nodesChunks.first().first() == nodesChunks.last().last()) {
+            val lastChunk = nodesChunks.removeAt(nodesChunks.lastIndex)
+            lastChunk.removeAt(lastChunk.lastIndex)
+            nodesChunks.first().addAll(0, lastChunk)
+        }
+
+        val indexOfChunkToKeep = nodesChunks.indexOfMaxBy { it.size }
+        val tags = originalWay.tags?.toMap()
+        var newWayId = -1L
+        return nodesChunks.mapIndexed { index, nodes ->
+            // TODO modification aware shit
+            if(index == indexOfChunkToKeep) OsmWay(originalWay.id, originalWay.version, nodes, tags)
+            else                            OsmWay(newWayId--, 0, nodes, tags)
+        }
+    }
+
+    /** Returns the elements that have been changed */
+    private fun updateRelations(originalWay: Way, newWays: List<Way>) : Collection<Relation> {
+        val relations = originalWay.fetchParentRelations()
+        val result = mutableSetOf<Relation>()
+        for (relation in relations) {
+            for(i in relation.members.size - 1 downTo 0) {
+                val relationMember = relation.members[i]
+                if (relationMember.type == WAY && relationMember.ref == originalWay.id) {
+                    if (!updateSpecialRelation(relation, i, newWays)) {
+                        updateNormalRelation(relation, i, originalWay, newWays)
+                    }
+                    result.add(relation)
+                }
+            }
+        }
+        return result
+    }
+
+    /** Returns whether it has been treated as a special relation type */
+    private fun updateSpecialRelation(relation: Relation, indexOfWayInRelation: Int, newWays: List<Way>): Boolean {
+        val relationType = relation.tags?.get("type") ?: ""
+        if (relationType == "restriction" || relationType == "destination_sign") {
```

Comment:
https://github.com/openstreetmap/iD/issues/6221 proposes to treat in this way all restrictions with `from` and `to` members. This seems to make sense.

Reply:
That sounds reasonable, but required also is a "via" node. And for `destination_sign`, the wiki says, that the role should actually be "intersection" (or "sign") and not "via".

So, depending on the relation type, it must still be decided which one is the via node/way. Not sure how iD handles this.